protocols/rendezvous/src/client.rs

```diff
@@ -314,73 +335,96 @@ impl NetworkBehaviour for Behaviour {
-                                    ))
-                                    .await;
-
-                                    (registration.record.peer_id(), registration.namespace)
-                                }
-                                .boxed()
-                            }));
-
-                        Event::Discovered {
-                            rendezvous_node,
-                            registrations,
-                            cookie,
+                if let Some((rendezvous_node, ns)) = self.waiting_for_discovery.remove(request_id) {
+                    let res = match response {
+                        Ok((registrations, cookie)) => {
+                            self.discovered_peers.extend(registrations.iter().map(
+                                |registration| {
+                                    let peer_id = registration.record.peer_id();
+                                    let namespace = registration.namespace.clone();
+
+                                    let addresses = registration.record.addresses().to_vec();
+
+                                    ((peer_id, namespace), addresses)
+                                },
+                            ));
+
+                            self.expiring_registrations
+                                .extend(registrations.iter().cloned().map(|registration| {
+                                    async move {
+                                        // if the timer errors we consider it expired
+                                        futures_timer::Delay::new(Duration::from_secs(
+                                            registration.ttl,
+                                        ))
+                                        .await;
+
+                                        (registration.record.peer_id(), registration.namespace)
+                                    }
+                                    .boxed()
+                                }));
+
+                            Event::Discovered {
+                                rendezvous_node,
+                                registrations,
+                                cookie,
+                            }
                         }
-                    }
-                    Err(error_code) => Event::DiscoverFailed {
-                        rendezvous_node,
-                        namespace: ns,
-                        error: error_code,
-                    },
-                };
+                        Err(error_code) => Event::DiscoverFailed {
+                            rendezvous_node,
+                            namespace: ns,
+                            error: error_code,
+                        },
+                    };
+
+                    return Some(res);
+                }
 
-                res
+                None
             }
-            _ => unreachable!(),
+            _ => unreachable!("rendezvous clients never receive requests"),
```

Comment:
This isn't safe. We cannot statically ensure that we don't receive a request, the remote could send us any message.

Reply:
In follow-up work, we could redesign this to have two enums: one for requests and one for responses. Then our codec could ensure that and inbound message on the client is always a response.